### PR TITLE
Enable VM live update rollout strategy always

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This PR enables the KubeVirt VM rollout strategy LiveUpdate, allowing changes to VM specs to dynamically propagate to their VMI.
Rollout strategy LiveUpdate is enabled using `VMLiveUpdateFeatures` Feature Gate and not a configuration knob. 
This because, from Kubevirt side, the feature is still protected by a FG, therefore it is still considered not stable in all the contexts and can potentially 1) be dropped, 2) changed in a breaking manner.
So the FeatureGate is still there and no configuration knob is exposed.
Nevertheless, according to the results of our tests, we can consider it as suitable for this use case and so we can handle it as part of the HCO opinionated deployment.
Eventual breaking changes will be evaluated and handled on the HCO side when consuming a future version of Kubevirt.

Kubevirt test_id:6867 is currently broken,
it will be fixed with https://github.com/kubevirt/kubevirt/pull/12073
It's not a blocker/critical issue so we can temporarely tolerate it.
The test will be enabled back when we will be able to consume a
newer and fixed Kubevirt release.

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-36370
```

**Release note**:

```release-note
KubeVirt VM rollout strategy LiveUpdate is now always enabled
```

### CODE REVIEW

**Review by tiraboschi — APPROVED** [review]

**Review by EdDev — COMMENTED** [review]
> KubeVirt VM rollout strategy LiveUpdate is now always enabled

I was under the impression that this feature is not yet GA on Kubevirt.
(otherwise you would not need to enable anything in HCO)

Could you please elaborate how can we commit to enable a non graduated feature by default?
Is there a policy that is followed here?

**Comment by tiraboschi** [comment]
> Could you please elaborate how can we commit to enable a non graduated feature by default? Is there a policy that is followed here?

As for https://github.com/kubevirt/kubevirt/blob/39dd8b6948de11ba7ccbda50fac711cb88421768/pkg/virt-config/feature-gates.go#L25

Feature gates on the Kubevirt are `intended for determining whether an optional feature is enabled or not at the cluster-level.` 

This is not 100% consistent with https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages but this is what we have until https://github.com/kubevirt/community/pull/251 will be accpeted.

**Comment by EdDev** [comment]
> > Could you please elaborate how can we commit to enable a non graduated feature by default? Is there a policy that is followed here?
> 
> As for https://github.com/kubevirt/kubevirt/blob/39dd8b6948de11ba7ccbda50fac711cb88421768/pkg/virt-config/feature-gates.go#L25
> 
> Feature gates on the Kubevirt are `intended for determining whether an optional feature is enabled or not at the cluster-level.`
> 
> This is not 100% consistent with https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages but this is what we have until [kubevirt/community#251](https://github.com/kubevirt/community/pull/251) will be accpeted.

I am just raising a flag that your interpretation of this is not how FGs are regarded per my understanding, including in Kubevirt.
From Kubevirt side, the feature is protected by a FG, therefore it is considered unstable and can potentially 1) be dropped, 2) changed in a breaking manner.
If the owners of the feature and the project maintainers want to say that it will not be dropped and not break compatibility backwards, they can and should remove the FG now, towards v1.3.

As long as the risks are clear, we should all be good.

**Comment by tiraboschi** [comment]
> I am just raising a flag that your interpretation of this is not how FGs are regarded per my understanding, including in Kubevirt. From Kubevirt side, the feature is protected by a FG, therefore it is considered unstable and can potentially 1) be dropped, 2) changed in a breaking manner. If the owners of the feature and the project maintainers want to say that it will not be dropped and not break compatibility backwards, they can and should remove the FG now, towards v1.3.

I'm not interpreting anything,
I'm simply reading this line: [This module (`feature-gates.go`) is intended for determining whether an optional feature is enabled or not at the cluster-level.](https://github.com/kubevirt/kubevirt/blob/39dd8b6948de11ba7ccbda50fac711cb88421768/pkg/virt-config/feature-gates.go#L25)

It's simply stating that those features, including `VMLiveUpdateFeaturesGate`, are optional and can be enabled or not at the cluster-level.
Nothing in that file, nor in the Release Note message nor in the PR description is currently suggesting me that `VMLiveUpdateFeaturesGate` is not GA and I can only take it at my own risk.

**Comment by EdDev** [comment]
> It's simply stating that those features, including `VMLiveUpdateFeaturesGate`, are optional and can be enabled or not at the cluster-level.
> Nothing in that file, nor in the Release Note message nor in the PR description is currently suggesting me that `VMLiveUpdateFeaturesGate` is not GA and I can only take it at my own risk.

Does this help?
https://kubevirt.io/user-guide/operations/activating_feature_gates/

File: controllers/operands/kubevirt.go
Comment (nunnatsa) [inline controllers/operands/kubevirt.go:455]: Please use ptr, and remove the `rolloutStrategy` var; e.g. 
```suggestion
		VMRolloutStrategy:            ptr.To(kubevirtcorev1.VMRolloutStrategyLiveUpdate)
```

File: controllers/operands/kubevirt_test.go
Comment (jean-edouard) [reply controllers/operands/kubevirt_test.go]: That's great, didn't know about `HaveValue()`, thanks!
Pushed.

**Comment by nunnatsa** [comment]
Missed the pushback from @EdDev 

/lgtm cancel

**Comment by nunnatsa** [comment]
Hi @jean-edouard.

We can ignore most of the failing lanes, but hco-e2e-kv-smoke-gcp lane should pass, but for this PR, it constantly failing, with the same errors. This is a subset of KV e2e. please take a look - it seems related to the PR itself.

**Comment by jean-edouard** [comment]
> Hi @jean-edouard.
> 
> We can ignore most of the failing lanes, but hco-e2e-kv-smoke-gcp lane should pass, but for this PR, it constantly failing, with the same errors. This is a subset of KV e2e. please take a look - it seems related to the PR itself.

Thank you for the heads-up @nunnatsa ! It's not immediately obvious to me why these tests are failing, I'll look into the failures as soon as I have some time.

**Comment by jean-edouard** [comment]
So... These tests are indeed incompatible with the LiveUpdate rollout strategy!
KubeVirt doesn't (yet) run with LiveUpdate enabled by default, which is why the tests are succeeding there.
We'll have to figure something out for those tests, I'll get back to you on that.
/hold

**Comment by acardace** [comment]
The instance type test fails but on the latest KubeVirt 1.2 it should pass as https://github.com/kubevirt/kubevirt/pull/11966 has been merged. @tiraboschi I guess this cannot pass because  `ci/prow/hco-e2e-kv-smoke-azure` installs kubevirt from the latest released commit but not from the latest merged commit in 1.2, is there way we can fix this?

**Comment by fabiand** [comment]
Hey. What is the status of this pr?

**Comment by nunnatsa** [comment]
> Hey. What is the status of this pr?

I think we're waiting for #2977

**Comment by EdDev** [comment]
> > Hey. What is the status of this pr?
> 
> I think we're waiting for #2977

Does it mean an explicit decision is taken to override the FG semantics?
If so, it will be great to see it written on the PR description.

**Comment by acardace** [comment]
> > > Hey. What is the status of this pr?
> > 
> > 
> > I think we're waiting for #2977
> 
> Does it mean an explicit decision is taken to override the FG semantics? If so, it will be great to see it written on the PR description.

Hey @EdDev , what do you mean exactly?

**Comment by EdDev** [comment]
> > > > Hey. What is the status of this pr?
> > > 
> > > 
> > > I think we're waiting for #2977
> > 
> > 
> > Does it mean an explicit decision is taken to override the FG semantics? If so, it will be great to see it written on the PR description.
> 
> Hey @EdDev , what do you mean exactly?

It was discussed earlier in this PR. See https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2846#issuecomment-2034896034

**Comment by tiraboschi** [comment]
/hold
waiting for https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2985 first

**Comment by tiraboschi** [comment]
hco-e2e-consecutive-operator-sdk-upgrades-azure lane succeeded.
ci/prow/hco-e2e-consecutive-operator-sdk-upgrades-aws

hco-e2e-operator-sdk-azure lane succeeded.
ci/prow/hco-e2e-operator-sdk-aws

hco-e2e-operator-sdk-sno-azure lane succeeded.
ci/prow/hco-e2e-operator-sdk-sno-aws

hco-e2e-upgrade-operator-sdk-azure lane succeeded.
ci/prow/hco-e2e-upgrade-operator-sdk-aws

hco-e2e-upgrade-operator-sdk-sno-azure lane succeeded.
ci/prow/hco-e2e-upgrade-operator-sdk-sno-aws

hco-e2e-upgrade-prev-operator-sdk-sno-azure lane succeeded.
ci/prow/hco-e2e-upgrade-prev-operator-sdk-sno-aws

**Comment by tiraboschi** [comment]
hco-e2e-consecutive-operator-sdk-upgrades-azure lane succeeded.
/override ci/prow/hco-e2e-consecutive-operator-sdk-upgrades-aws

hco-e2e-operator-sdk-azure lane succeeded.
/override ci/prow/hco-e2e-operator-sdk-aws

hco-e2e-operator-sdk-sno-azure lane succeeded.
/override ci/prow/hco-e2e-operator-sdk-sno-aws

hco-e2e-upgrade-operator-sdk-azure lane succeeded.
/override ci/prow/hco-e2e-upgrade-operator-sdk-aws

hco-e2e-upgrade-operator-sdk-sno-azure lane succeeded.
/override ci/prow/hco-e2e-upgrade-operator-sdk-sno-aws

hco-e2e-upgrade-prev-operator-sdk-sno-azure lane succeeded.
/override ci/prow/hco-e2e-upgrade-prev-operator-sdk-sno-aws

**Pull request merged**
